crates/server/src/store/file/authority.rs

```diff
@@ -54,6 +58,66 @@ impl FileAuthority {
         InMemoryAuthority::new(origin, records, zone_type, allow_axfr).map(Self)
     }
 
+    /// Read given file line by line and recursively invokes reader for
+    /// $INCLUDE directives
+    ///
+    /// TODO: it looks hacky as far we effectively duplicate parser's functionallity
+    /// (at least partially) and performing lexing twice.
+    /// Better solution requires us to change lexer to deal
+    /// with Lines-like iterator instead of String buf (or capability to combine a few
+    /// lexer instances into a single lexer).
+    ///
+    /// TODO: we should probably track depth of recursion here and set max level (or
+    /// even make it configurable, in practice it's hard to expect INCLUDES deeper than
+    /// 3-4 levels anyways).
+    ///
+    /// TODO: $INCLUDE could specify domain name -- to support on-flight swap for Origin
+    /// value we definitely need to rethink and rework loader/parser/lexer
+    fn read_file(zone_path: PathBuf, buf: &mut String) -> Result<(), String> {
+        let file = File::open(&zone_path)
+            .map_err(|e| format!("failed to read {}: {:?}", zone_path.display(), e))?;
+        let reader = BufReader::new(file);
+        for line in reader.lines() {
+            let content = line.unwrap();
+            let mut lexer = Lexer::new(&content);
+
+            match (lexer.next_token(), lexer.next_token()) {
+                (Ok(Some(Token::Include)), Ok(Some(Token::CharData(include_path)))) => {
+                    // RFC1035 (section 5) does not specify how filename for $INCLUDE
+                    // should be resolved into file path. The underlying code implements the
+                    // following:
+                    // * if the path is absolute (relies on Path::is_absolute), it uses normalized path
+                    // * otherwise, it joins the path with parent root of the current file
+                    //
+                    // TODO: Inlining files specified using non-relative path might potentially introduce
+                    // security issue in some cases (e.g. when working with zone files from untrusted sources)
+                    // and should probably be configurable by user.
+                    let include_path = Path::new(&include_path);
+                    let include_zone_path = if include_path.is_absolute() {
+                        include_path.to_path_buf()
+                    } else {
+                        let parent_dir =
+                            zone_path.parent().expect("file has to have parent folder");
+                        parent_dir.join(include_path)
+                    };
+
+                    let mut include_buf = String::new();
+
+                    debug!("loading included zone file: {:?}", include_zone_path);
```

Comment:
I think this should be an `info!` rather than debug. And maybe a little more information, i.e. "including file {include_zone_path} into {zone}", or something along those lines.